extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+
+      for (final StageDefinition stageDef : queryDef.getStageDefinitions()) {
+        if (stageDef.equals(finalShuffleStageDef)) {
+          builder.add(StageDefinition.builder(stageDef).shuffleCheckHasMultipleValues(true));
+        } else {
+          builder.add(StageDefinition.builder(stageDef));
+        }
+      }
+
+      // Then, add a segment-generation stage.
+      final DataSchema dataSchema = generateDataSchema(querySpec, querySignature, queryClusterBy, columnMappings);
+      builder.add(
+          StageDefinition.builder(queryDef.getNextStageNumber())
+                         .inputs(new StageInputSpec(queryDef.getFinalStageDefinition().getStageNumber()))
+                         .maxWorkerCount(tuningConfig.getMaxNumWorkers())
+                         .processorFactory(
+                             new SegmentGeneratorFrameProcessorFactory(
+                                 dataSchema,
+                                 columnMappings,
+                                 tuningConfig
+                             )
+                         )
+      );
+
+      return builder.build();
+    } else if (querySpec.getDestination() instanceof TaskReportMSQDestination) {
+      return queryDef;
+    } else {
+      throw new ISE("Unsupported destination [%s]", querySpec.getDestination());
+    }
+  }
+
+  private static DataSchema generateDataSchema(
+      MSQSpec querySpec,
+      RowSignature querySignature,
+      ClusterBy queryClusterBy,
+      ColumnMappings columnMappings
+  )
+  {
+    final DataSourceMSQDestination destination = (DataSourceMSQDestination) querySpec.getDestination();
+    final boolean isRollupQ = isRollupQuery(querySpec.getQuery());
+
+    final Pair<List<DimensionSchema>, List<AggregatorFactory>> dimensionsAndAggregators =
+        makeDimensionsAndAggregatorsForIngestion(
+            querySignature,
+            queryClusterBy,
+            destination.getSegmentSortOrder(),
+            columnMappings,
+            isRollupQ,
+            querySpec.getQuery()
+        );
+
+    return new DataSchema(
+        destination.getDataSource(),
+        new TimestampSpec(ColumnHolder.TIME_COLUMN_NAME, "millis", null),
+        new DimensionsSpec(dimensionsAndAggregators.lhs),
+        dimensionsAndAggregators.rhs.toArray(new AggregatorFactory[0]),
+        makeGranularitySpecForIngestion(isRollupQ, querySpec.getQuery()),
+        new TransformSpec(null, Collections.emptyList())
+    );
+  }
+
+  private static GranularitySpec makeGranularitySpecForIngestion(boolean isRollupQ, Query<?> query)
+  {
+    if (!isRollupQ) {
+      return new ArbitraryGranularitySpec(Granularities.NONE, false, Intervals.ONLY_ETERNITY);
+    } else {
+      final String queryGranularity = query.getContextValue(GroupByQuery.CTX_TIMESTAMP_RESULT_FIELD_GRANULARITY, "");
+
+      if (checkIfTimeColumnsAreEqual((GroupByQuery) query) && !queryGranularity.isEmpty()) {
```

Comment:
Nit: `checkIfTimeColumnsAreEqual` -> `timeColumnsAreEqual`. Makes the `if` read a bit more smoothly.

Reply:
Works for me.

Reply:
I renamed it to `timeIsGroupByDimension`, which is a more accurate name given how it works. I also added this comment describing the logic:

```
  /**
   * Checks that a {@link GroupByQuery} is grouping on the primary time column.
   *
   * The logic here is roundabout. First, we check which column in the {@link GroupByQuery} corresponds to the
   * output column {@link ColumnHolder#TIME_COLUMN_NAME}, using our {@link ColumnMappings}. Then, we check for the
   * presence of an optimization done in {@link DruidQuery#toGroupByQuery()}, where the context parameter
   * {@link GroupByQuery#CTX_TIMESTAMP_RESULT_FIELD} and various related parameters are set when one of the dimensions
   * is detected to be a time-floor. Finally, we check that the name of that dimension, and the name of our time field
   * from {@link ColumnMappings}, are the same.
   */
```